crates/plugins/mediator-coordination/src/forward/routing.rs

```diff
@@ -0,0 +1,252 @@
+use std::sync::Arc;
+
+use axum::response::{IntoResponse, Response};
+
+use database::Repository;
+use didcomm::{AttachmentData, Message};
+use hyper::StatusCode;
+use mongodb::bson::doc;
+use serde_json::{json, Value};
+
+use crate::{
+    model::stateful::entity::{Connection, RoutedMessage},
+    web::{error::MediationError, AppState, AppStateRepository},
+};
+
+use super::error::RoutingError;
+
+/// Mediator receives forwarded messages, extract the next field in the message body, and the attachments in the message
+/// then stores the attachment with the next field as key for pickup
+pub async fn mediator_forward_process(
+    state: &AppState,
+    payload: Message,
+) -> Result<Message, Response> {
+  
+    let result = handler(state, payload).await.unwrap();
+    Ok(result)
+}
+
+async fn checks(
+    message: &Message,
+    connection_repository: &Arc<dyn Repository<Connection>>,
+) -> Result<String, Response> {
+    let next = message.body.get("next").and_then(Value::as_str);
+    match next {
+        Some(next) => next,
+        None => {
+            let response = (StatusCode::BAD_REQUEST, RoutingError::MalformedBody.json());
+            return Err(response.into_response());
+        }
+    };
+    // Check if the client's did in mediator's keylist
+    let _connection = match connection_repository
+        .find_one_by(doc! {"keylist": doc!{ "$elemMatch": { "$eq": &next}}})
+        .await
+        .map_err(|_| MediationError::RepostitoryError)
+        .unwrap()
+    {
+        Some(connection) => connection,
+        None => {
+            let response = (
+                StatusCode::UNAUTHORIZED,
+                MediationError::UncoordinatedSender.json(),
+            );
+            return Err(response.into_response());
+        }
+    };
+    Ok(next.unwrap().to_string())
+}
+
+async fn handler(state: &AppState, message: Message) -> Result<Message, MediationError> {
+    let AppStateRepository {
+        message_repository,
+        connection_repository,
+        ..
+    } = state
+        .repository
+        .as_ref()
+        .ok_or_else(|| MediationError::RepostitoryError)?;
+    let next = checks(&message, connection_repository).await.unwrap();
+
+    let attachments = message.attachments.unwrap_or_default();
+    for attachment in attachments {
+        let attached = match attachment.data {
+            AttachmentData::Json { value: data } => data.json,
+            AttachmentData::Base64 { value: data } => json!(data.base64),
+            AttachmentData::Links { value: data } => json!(data.links),
+        };
+        message_repository
+            .store(RoutedMessage {
+                id: None,
+                message: json!(attached),
```

Comment:
Be careful. The attachment should already be in Json format here.
```suggestion
                message: attached,
```